decoder/vaapidecoder_h265.cpp

```diff
@@ -0,0 +1,1784 @@
+    COPY_FIELD (sps, log2_diff_max_min_pcm_luma_coding_block_size);
+    COPY_FIELD (sps, max_transform_hierarchy_depth_intra);
+    COPY_FIELD (sps, max_transform_hierarchy_depth_inter);
+    COPY_FIELD (pps, init_qp_minus26);
+    COPY_FIELD (pps, diff_cu_qp_delta_depth);
+    pic_param->pps_cb_qp_offset = pps->cb_qp_offset;
+    pic_param->pps_cr_qp_offset = pps->cr_qp_offset;
+    COPY_FIELD (pps, log2_parallel_merge_level_minus2);
+    COPY_FIELD (pps, num_tile_columns_minus1);
+    COPY_FIELD (pps, num_tile_rows_minus1);
+    for (i = 0; i <= pps->num_tile_columns_minus1; i++)
+        pic_param->column_width_minus1[i] = pps->column_width_minus1[i];
+    for (; i < 19; i++)
+        pic_param->column_width_minus1[i] = 0;
+    for (i = 0; i <= pps->num_tile_rows_minus1; i++)
+        pic_param->row_height_minus1[i] = pps->row_height_minus1[i];
+    for (; i < 21; i++)
+        pic_param->row_height_minus1[i] = 0;
+
+    COPY_BFM (slice_parsing_fields, pps, lists_modification_present_flag);
+    COPY_BFM (slice_parsing_fields, sps, long_term_ref_pics_present_flag);
+    pic_param->slice_parsing_fields.bits.sps_temporal_mvp_enabled_flag =
+    sps->temporal_mvp_enabled_flag;
+    COPY_BFM (slice_parsing_fields, pps, cabac_init_present_flag);
+    COPY_BFM (slice_parsing_fields, pps, output_flag_present_flag);
+    COPY_BFM (slice_parsing_fields, pps, dependent_slice_segments_enabled_flag);
+    pic_param->slice_parsing_fields.
+    bits.pps_slice_chroma_qp_offsets_present_flag =
+    pps->slice_chroma_qp_offsets_present_flag;
+    COPY_BFM (slice_parsing_fields, sps, sample_adaptive_offset_enabled_flag);
+    COPY_BFM (slice_parsing_fields, pps, deblocking_filter_override_enabled_flag);
+    pic_param->slice_parsing_fields.bits.pps_disable_deblocking_filter_flag =
+    pps->deblocking_filter_disabled_flag;
+    COPY_BFM (slice_parsing_fields, pps,
+    slice_segment_header_extension_present_flag);
+    pic_param->slice_parsing_fields.bits.RapPicFlag = picture->m_RapPicFlag;
+    pic_param->slice_parsing_fields.bits.IdrPicFlag = VAAPI_PICTURE_FLAG_IS_SET (picture, VAAPI_PICTURE_FLAG_IDR);
+    pic_param->slice_parsing_fields.bits.IntraPicFlag = picture->m_IntraPicFlag;
+
+    COPY_FIELD (sps, log2_max_pic_order_cnt_lsb_minus4);
+    COPY_FIELD (sps, num_short_term_ref_pic_sets);
+    pic_param->num_long_term_ref_pic_sps = sps->num_long_term_ref_pics_sps;
+    COPY_FIELD (pps, num_ref_idx_l0_default_active_minus1);
+    COPY_FIELD (pps, num_ref_idx_l1_default_active_minus1);
+    pic_param->pps_beta_offset_div2 = pps->beta_offset_div2;
+    pic_param->pps_tc_offset_div2 = pps->tc_offset_div2;
+    COPY_FIELD (pps, num_extra_slice_header_bits);
+
+    /*Fixme: Set correct value as mentioned in va_dec_hevc.h */
+    pic_param->st_rps_bits = 0;
+
+    return true;
+}
+
+void VaapiDecoderH265::fill_iq_matrix_4x4 (VAIQMatrixBufferHEVC * iq_matrix,
+    H265ScalingList * scaling_list)
+{
```

Comment:
4x4, 8x8  16x16 32x32 all seems same function, can we use a common function, pass array and zigzag function pointer to it?